File: crux-core/src/crux/api/QueryState.java
package crux.api;

import clojure.lang.Keyword;

import java.util.Date;
import java.util.Map;

public class QueryState {
    public enum QueryStatus {
        IN_PROGRESS, COMPLETED, FAILED
    }

    public static class QueryError {
        public final String errorClass;
        public final String errorMessage;

        public QueryError(String errorClass, String errorMessage) {
            this.errorClass = errorClass;
            this.errorMessage = errorMessage;
        }
    }

    public final Object queryId;
    public final Date startedAt;
    public final Date finishedAt;
    public final QueryStatus status;
    public final Map<Keyword, ?> query;
    public final QueryError error;

    public QueryState(Object queryId, Date startedAt, Date finishedAt, QueryStatus status, Map<Keyword, ?> query, QueryError error) {
        this.queryId = queryId;
        this.startedAt = startedAt;
        this.finishedAt = finishedAt;
        this.status = status;
        this.query = query;
        this.error = error;
    }
}

File: core/src/crux/query_state.clj
(ns ^:no-doc crux.query-state
  (:import (crux.api IQueryState IQueryState$IQueryError IQueryState$QueryStatus)
           (java.io Writer)))

(defn ->query-status [status]
  (case status
    :failed IQueryState$QueryStatus/FAILED
    :completed IQueryState$QueryStatus/COMPLETED
    :in-progress IQueryState$QueryStatus/IN_PROGRESS))

(defrecord QueryError [type message]
  IQueryState$IQueryError
  (getErrorClass [this] type)
  (getErrorMessage [this] message))

(defrecord QueryState [query-id started-at finished-at status query error]
  IQueryState
  (getQueryId [this] query-id)
  (getStartedAt [this] started-at)
  (getFinishedAt [this] finished-at)
  (getStatus [this] (->query-status status))
  (getQuery [this] query)
  (getError [this] error))

(defmethod print-method QueryState [qs ^Writer w]
  (.write w "#xt/query-state ")
  (print-method (into {} qs) w))

(defmethod print-method QueryError [qs ^Writer w]
  (.write w "#xt/query-error ")
  (print-method (into {} qs) w))

(defn ->QueryError [error]
  (let [{:keys [type message]} error]
    (QueryError. type message)))

(defn ->QueryState [{:keys [query-id started-at finished-at status error query] :as query-state}]
  (QueryState. query-id
               started-at
               finished-at
               status
               query
               (when error
                 (->QueryError error))))
